subsys/bluetooth/common/rpa.c

```diff
@@ -14,20 +14,41 @@
 #include <stddef.h>
 #include <errno.h>
 #include <string.h>
-#include <sys/atomic.h>
-#include <sys/util.h>
-#include <sys/byteorder.h>
-#include <debug/stack.h>
-
-#include <tinycrypt/constants.h>
-#include <tinycrypt/aes.h>
-#include <tinycrypt/utils.h>
-#include <tinycrypt/cmac_mode.h>
 
 #define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_RPA)
 #define LOG_MODULE_NAME bt_rpa
 #include "common/log.h"
 
+#include <bluetooth/crypto.h>
+
+#if defined(CONFIG_BT_CTLR) && defined(CONFIG_BT_HOST_CRYPTO)
+#include "../controller/util/util.h"
+#include "../controller/hal/ecb.h"
+#endif /* defined(CONFIG_BT_CTLR) && defined(CONFIG_BT_HOST_CRYPTO) */
+
+static int internal_rand(void *buf, size_t len)
+{
+/* Force using controller rand function. */
+#if defined(CONFIG_BT_CTLR) && defined(CONFIG_BT_HOST_CRYPTO)
+	return util_rand(buf, len);
+#else
+	return bt_rand(buf, len);
+#endif
+}
+
+static int internal_encrypt_le(const u8_t key[16], const u8_t plaintext[16],
+			       u8_t enc_data[16])
+{
+/* Force using controller encrypt function if supported. */
+#if defined(CONFIG_BT_CTLR) && defined(CONFIG_BT_HOST_CRYPTO) && \
+    defined(CONFIG_BT_CTLR_LE_ENC_SUPPORT)
```

Comment:
BT_CTLR_LE_ENC_SUPPORT is not documented at all (in fact none of the BT_CTLR_*_SUPPORT macros).
ecb.c is not conditionally compiled based on BT_CTLR_LE_ENC_SUPPORT.

I understand your intention, but we should clarify the purpose of BT_CTLR_*_SUPPORT macros

Reply:
@George-Stefan Yeah, I just took the first one I found :) Is using BT_CTLR_LE_ENC better?
The _SUPPORT ones are meant to be  internal ones to make Kconfig dependencies easier.

Reply:
The LE Encrypt command is not conditionally compiled out by anything in the controller. Should it perhaps? I see you added empty functions instead. I think all other calls to ecb_encrypt or ecb_encrypt_le are behind guards for BT_CTLR_LE_ENC .

Reply:
There is a mix of implicit and explicit capabilities the relation of which I don't quite understand:
- functions from ecb.h
- BT_HOST_CRYPTO
- BT_CTLR_CRYPTO
- BT_CTLR_LE_ENC_SUPPORT
- BT_CTLR_PRIVACY_SUPPORT

Reply:
BT_HOST_CRYPTO -> compile in host/crypto.c for bt_rand/bt_encrypt_le/bt_encrypt_be
BT_CTLR_CRYPTO -> compile in controller/crypto.c for bt_rand/bt_encrypt_le/bt_encrypt_be
The first two should be mutually exlusive. Which is why HOST_CRYPTO is set automatically if not CTLR_CRYPTo

BT_CTLR_LE_ENC_SUPPORT -> The controller can support the LE Encryption command
BT_CTLR_PRIVACY_SUPPORT -> The controller can support all related LE controller-based privacy commands.

BT_CTLR_LE_ENC -> Enable support for LE encryption (depends on controller support).
BT_CTLR_PRIVACY -> Enable support for LE controller-based privacy (depends on controller support).

More generally BT_CTLR_<feature>_SUPPORT is set by the different controllers to indicate that this feature can be enabled.
BT_CTLR_<feature> is visible if the selected controller has _SUPPORT, and the user can enable or disable this feature.

Reply:
The host depends on the bt_rand/bt_encrypt_le/bt_encrypt_be functions.

Reply:
Switched to BT_CTLR_LE_ENC.
@George-Stefan @cvinayak Please comment if this is wrong.

Reply:
I believe that BT_CTLR_CRYPTO should be conditioned by a BT_CTLR_CRYPTO_SUPPORT as well.
Now, there is the implicit assumption that the controller has crypto acceleration/support.

And because of this, there is this confusion where you disable BT_CTLR_LE_ENC_SUPPORT/BT_CTLR_PRIVACY_SUPPORT but BT_CTLR_CRYPTO is enabled by default.

I can see how you could have BT_CTLR_LE_ENC_SUPPORT/BT_CTLR_PRIVACY_SUPPORT without BT_CTLR_CRYPTO.
So, internal_encrypt_le() is not needed.
What do you think?

Reply:
@George-Stefan You mean introduce a new Kconfig support flag? I like that suggestion.

@cvinayak Agree to add BT_CTLR_CRYPTO_SUPPORT, and have BT_CTLR_CRYPTO depend on this flag?

Reply:
> @George-Stefan You mean introduce a new Kconfig support flag? I like that suggestion.

Yes

Reply:
@George-Stefan I agree that that could be done. But I'd like to do that separately. So merge this PR, and then clean up the CTLR_CRYPTO later.

Reply:
@George-Stefan I agree with @joerchan. This PR fixes a real bug, and the shuffling of Kconfig options can be done as a separate Pull Request where we can all focus on the cleanup.

Reply:
ok